### PR TITLE
Event Tracing API C# Bindings

### PR DESCRIPTION
**Pull request by seanjparker**
For [UTY-2671](https://improbableio.atlassian.net/browse/UTY-2671)

#### Description
Initial work for the implementation of the Event Tracing API, introduces some internal C# bindings for the C API

#### Tests
How did you test these changes prior to submitting this pull request?
What automated tests are included in this PR?
WIP

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?
WIP

- [x] Changelog

### CODE REVIEW

File: workers/unity/Packages/io.improbable.worker.sdk/CIO.cs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+namespace Improbable.Gdk
+{
+    internal unsafe class CIO
+    {
+#if UNITY_IOS
+        private const string WorkerLibrary = "__Internal";
+#else
+        private const string WorkerLibrary = "improbable_worker";
+#endif
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.worker.sdk/CIO.cs:18]: We could put this in its own file and reference it both here and in `CEventTrace.cs`

File: workers/unity/Packages/io.improbable.worker.sdk/CIO.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
+using System.Runtime.InteropServices;
+using Improbable.Worker.CInterop.Internal;
+using Int64 = System.Int64;
+using Uint64 = System.UInt64;
+using Uint32 = System.UInt32;
+using Uint8 = System.Byte;
+using Char = System.Char;
+using IntPtr = System.IntPtr;
+
+namespace Improbable.Gdk
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.worker.sdk/CIO.cs:10]: I'd possible use the same namespace as the worker SDK to minimize changes required in the future. (`Improbable.Worker.CInterop.Internal` I think)

Comment (jamiebrynes7) [reply workers/unity/Packages/io.improbable.worker.sdk/CIO.cs:10]: Same thing applies for `CEventTrace.cs`

File: workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+            public Uint8 ItemType;
+
+            [StructLayout(LayoutKind.Explicit)]
+            public struct Union
+            {
+                [FieldOffset(0)] public Span Span;
+
+                [FieldOffset(0)] public Event Event;
+            }
+        }
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs:173]: `Item` needs a member of type `Item.Union`. The header file both declares the union and has a field called `item`: 

```c
typedef struct Trace_Item {
  /** The type of the item, defined using Trace_ItemType. */
  uint8_t item_type;
  union Trace_Item_Union {
    Trace_Span span;
    Trace_Event event;
  } item;
} Trace_Item;
```

File: workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+         * returned a non-zero item size. Instead, this can be indirectly checked by passing the size to
+         * this method and checking its return value for an error.
+         *
+         * The caller is responsible for ensuring that the provided stream has sufficient remaining capacity
+         * to hold the serialized item.
+         *
+         * Returns 1 on success, 0 on error. Call Io_Storage_GetLastError to get the associated error
+         * message.
+         */
+        // TODO: Check where Io_Storage_GetLastError exists
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs:326]: TODO?

Comment (seanjparker) [reply workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs:326]: Oh right, I missed this one. This was before the `CIO.cs` file was written. The `GetLastError` function exists in that file now

Comment (seanjparker) [reply workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs:326]: After I created the `CIO.cs` file, I noticed that it doesn't provide a function signature for `Io_Storage_GetLastError`. It does provide a function called `Io_Stream_GetLastError` so I'm not sure if it's something that was removed or not implemented yet.

Comment (jamiebrynes7) [reply workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs:326]: cc @improbable-jonphillips @tbimprob this looks like a docs error?

Comment (tbimprob) [reply workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs:326]: Yes this was a documentation error (should be ` * Returns 1 on success, 0 on error. Call Trace_GetLastError to get the associated error`) it has been fixed from 14.7.1+ https://github.com/improbable/everything/pull/5459/files

File: workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs (hunk 3 above)
Comment (seanjparker) [reply workers/unity/Packages/io.improbable.worker.sdk/CEventTrace.cs]: Added the new field to the Item struct.

**Pull request merged**